### DISCUSSION TITLE
Internship_Report_Ananya

### DISCUSSION THREAD

**Discussion by ananyas168** [petermr/petermr#24] [Show and tell]

**Comment by ananyas168** [comment]
Hi, I am Ananya Singha, a 4th-year undergrad doing a Major in EECS from IISER Bhopal. I am an ML enhusiast. 

This discussion channel is my medium to talk about all the subprojects I worked on during this 3-month internship under the guidance of Dr. Peter Murray Rust.

**Reply by EmanuelFaria** [reply]
Hi Ananya, 

I finally got a chance to review your draft report and mark it up with some comments and questions. I find it cumbersome to do that within this discussion page so I've made a [google doc where you can see my comments, here](https://docs.google.com/document/d/17qGl4VcH_pdR91u9R6UJqmddcQV2uSto/edit?usp=sharing&ouid=110062979113982375503&rtpof=true&sd=true).

Keep in mind, my background is in writing for business and marketing, not science, so use my comments as tips only, and use your best judgment.

I hope this helps. :)

**Reply by EmanuelFaria** [reply]
Hi Ananya, forget the Google Doc. I've simplified my notes below. I hope they make sense and help! Let me know if you have questions about any of it. By the way, if you found any of this helpful, you may also get something from reading my notes on your teammates reports.

**Comment by ananyas168** [comment]
## **PROJECT 1**: Building tools for analysis of **Terpene Classification Task**:
primary goal: extracting biosynthetic pathways from images in scientific literature.
Let's first dive into some basics. And ask some crucial Questions. What terpene classification is and why are we doing terpene classification?
Let's begin with understanding what terpene is.

## Q. What is terpene?
Terpene are chemical compounds found in biosynthetic pathways. Terpenes are highly aromatic compounds that determine the smell of many plants and herbs, such as rosemary and lavender. In biosynthetic pathway there are two major pathway which result in terpene production, MVA Pathway and MEP Pathway. 
## Q. What is metabolic pathway?
In [biochemistry](https://en.wikipedia.org/wiki/Biochemistry), a metabolic pathway is a linked series of [chemical reactions](https://en.wikipedia.org/wiki/Chemical_reaction) occurring within a [cell](https://en.wikipedia.org/wiki/Cell_(biology)). The [reactants](https://en.wikipedia.org/wiki/Reactant), products, and intermediates of an enzymatic reaction are known as [metabolites](https://en.wikipedia.org/wiki/Metabolites), which are modified by a sequence of chemical reactions [catalyzed](https://en.wikipedia.org/wiki/Catalyze) by [enzymes](https://en.wikipedia.org/wiki/Enzyme).[[1]](https://en.wikipedia.org/wiki/Metabolic_pathway#cite_note-Nelson-1): 26  In most cases of a metabolic pathway, the [product](https://en.wikipedia.org/wiki/Product_(chemistry)) of one enzyme acts as the [substrate](https://en.wikipedia.org/wiki/Substrate_(chemistry)) for the next. However, side products are considered waste and removed from the cell.These enzymes often require dietary minerals, vitamins, and other [cofactors](https://en.wikipedia.org/wiki/Cofactor_(biochemistry)) to function

## Q  Why terpene pathway Classification?
Our Goal of terpene pathway image classification is a subset of a bigger project. To accomplish the task we collected images from Alex Pico's website and tried to classify those images with help human annotators. This subtask carried out by Human annotators gave us baseline standard for our higher level task of using Machine Learning to classify terpene pathway images captured over the internet.Further the statistics from the human annotation experiment made us realize how ambigous it gets for humans to distinguish Terpene pathway from Pre-Terpene Pathway. The analysis also pointed out that its very easy for human to capture low level classification task i.e to identify is the given image is pathway or not ot if the pathway image is metabolic pathway or not.

## Brief on the Experimentation (Terpene Classification by Human Annotators)
To sucessfully operate on terpene classification task we took dataset from a trusted source. In Alex's website you can more than 300 pathway images some belong to terpene and others not. Therefore it suited our purpose the best, as we wanted to understnd the statistics behind terpene classification task and learn how difficult is gets for human to classifiy images to narrower domains like [pre- terpene, terpene] from broader domain like [Pathway, Metabolic].
The table below will give you a glimse of our data looks like for the experimentation.

![image](https://user-images.githubusercontent.com/66965350/161371325-85c64dd7-f1ca-450d-8e9f-00ba8cb2c592.png)

The core idea is to undertand how easy or difficult it is to carry out terpene clasification by human annotators along with that collection labeled data for procuring training/ test data for machine Leraning Frame work. Therefore we circulated the 30 random iamges among 8 annotators coming from different background and asked to classify each of this image. A small sneak peak of the annotated data from one of the annotators in show in the fig below:

![image](https://user-images.githubusercontent.com/66965350/161371924-fc18a16b-58b7-456f-8fc3-51842b967859.png)

The annotated data from the 8 snnotators is then used for the statistical annalysis. You can find the code [here](https://github.com/petermr/CEVOpen/blob/master/statistical_ananlysis_2022/statistical_analysis_annotators.ipynb). You can find discription about the code below. Please check it out. It will take you to an interactive colab session where you can get hands on experience of our experimentation.

## Metrics used for Analysis
1. [Correlation coefficient](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.corr.html)
2. [Confusion Matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html)
3. [Cohens kappa](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html)

We found that the best tool to compare performances between annotator is best showcased by cohen's kappa for class labels Pre-Terpene and Terpene.
whereas Confusion matrix is best for Class labels Pathway and Metabolic. Therefore we use cohen's kappa and confusion matrix to compare and analyse the performances between annotators.


## Results and Observation:
1. The experiment suggested that humans are easily able to distinguish between pathway and non-pathway images. Suggesting that human indeed are good at boarder level task and the rules defined for the pathway label is well understood by the human. 
2. Similary the experiment suggested that human are capable of identifing Metabolic pathway easily, and have the least disagreement among annotators.
3. The highest disagreement lies for the class label terpene. This potrayed that it is veru difficult among huamns to distinguish between terpene and and non terpene pathway.
4. Though the human performance for pre-terpene class was better and there was less disagrement, which suggested that rules for preterpene well defined and human coming from different background are able to understand it. 

The table below gives information of interannotator agrement using cohens kappa. The higer the value more is the agrement between annotators.
![image](https://user-images.githubusercontent.com/66965350/161372970-09cc6e69-8b75-46ea-bfb2-ff6b428852aa.png)

From the above table we can see that the average value for terpene is below 0.5 where as average vakue for pre-terpene is 0.8. This suggest that there was high ambiguty among annotators regarding terpene classification. Whereas there was less ambiguity for pre terpene class suggesting humans are confident about pre-terpene pathways

## Future Take
1. Modify the rules for terpene images. So that the ambiguity among the annotators reduces.
2. Finding the troubling imges. And identify what the cause of major confusion among annotators.
  
## In the [statistical_analysis_2022](https://github.com/petermr/CEVOpen/tree/master/statistical_ananlysis_2022) folder you will find:
1. **df_generation.py** : script file that generates dataframe from all the annotators csv files for a concerned column name [ex. pathway, terpene]
2. **statistical_analysis.iypnb**: is an interactive jupyter notebook that computes the correlation between annotators based on different metrics. In order to see how we are comparing and analyzing the annotator's performance. Open the notebook in your google collaboratory by just clicking on the "open colab" button at the top of the notebook, and then experience our methodology for the analysis. 
3. The subfolder **example_data** gives a glimpse of what the original data will look like.

**Reply by EmanuelFaria** [reply]
PROJECT 1: Building tools for analysis of Terpene Classification Task:



> TITLE: If I understand your intention, I “think” your title here should more specifically describe what you’ve writing in the next section you’ve called “Purpose” (which you may want to change to “Objective”). If I’m not mistaken, general “Terpene Classification” may better describe the main project, of which the sub-project is Increment in its accomplishment (you’re working on a “tool” in the overall “toolkit” called docanalysis, I think). Maybe something like “Developing Tools for Machines to Classify Terpenes”?



> BACKGROUND: If I’ve understood the assignment so far, vetween “Project 1” and “Purpose”, consider adding a heading called “BACKGROUND”. Here you could start by explaining that what you’re writing about in this report, and how this sub-project you’ve participated in fits into the overall toolkit project as briefly as you can, and then link to the “Toolkit’s” description page, (which I think you and Roopa and Ananya will work on together and post separately).



> PURPOSE/OBJECTIVE: Here you could briefly explain that the project being described below is a key part of making a functional tool that will be part of a toolkit that will extract data from images… etc.. then go right to describing this project in the next section.

With the help of constant cmmunal disussion we are tryingto bild a machine learning toolkit that could identify terpenes from the images (source alex pico's wikipathway).



> Suggestions: 1) Link “wikipathway.org” to URL with best, shortest description and include “.org” in sentence. 2) Replace “could” with “will”. 3) Consider moving the part about “with the help of constant communal discussion“ to your “About Me” section and introduce yourself, the team and how you all worked remotely there so you can go right to the project here and in the (sub)sections below.

Our Purpose:



> Perhaps, “Objective” rather than Purpose?

We needed to come to a confirmation that whether the extracted images on wikipathway page from various papers, articles, journels etc. consistis of Terpene Pathway.



> This is good “We want..” statement.. consider adding a “So that… (outcome). for the purpose of… (being used by scientists to x without y?)

For this purpose we asked various people with any background to annotate the CSV file so that we could analyze the database and get the needed confirmation.



> I think this would be a great first sentence under “Methodology”. It might be useful to think about structuring each step something like: Do/did [x] so that we accomplish [y] (without [z] happening). Then you could replace “For this purpose” with “First,” and then list the steps.

Methodology:

With the collected raw data following the use of statistical analysis based on inter-annotator agreement and correlation coefficient we analyzed the annotation and came to a certain conclusion.



> Break this into at least two sentences. My layman brain got confused. ;)

Process:

To formlate a basic structure for construction of the tool, we started with creating a simple rule book that an annotator could go through and fill out the coloumns accordingly:



> This might be easier to explain as steps. Also I’m not sure if “construction” should be “development”, but Peter would know.

Intially the tabel to be filled looked like:



With the help of constant communal discussion we came with the draft of rules which we circulated across to get the inputs.



> Suggestion: Replace “we came with the draft of” with “our team drafted the following”. I don’t know what “circulated across to get the inputs” means.



To sucessfully operate on terpene classification task we took dataset from a trusted source. 



> … from a trusted source: wikipathways.org”



> This sentence feels to me like the beginning of a change of topic. If so, suggest you add a sub-sub-heading that defines to the reader what it is they’re about to read. 



> Here's an old tip for writing marketing copy that can apply in the different sections of this report and practically anything else you’ll write:



> 1. Tell them what you’re going to tell them.



> 2. Tell them.



> 3. Tell them what you told them.



> Another think I’ve learned the hard way is to always begin a paragraphs and sentences with the “WHY” before you get to the “WHAT” or “HOW”. Starting with the WHY gives the reader the context required to understand the significance of the WHAT and HOW. 



> You’ve sort of donE that here with “To successfully operate on terpene…”,, but I go into read the paragraph wondering WHY it’s important. So, maybe a stronger WHY could be something along the lines of, “It is important that our terpene classifications were x and not y, so we [did] “WHAT” by “HOW”.



> Here’s an example from my previous working in marketing real estate investments… there were about 4,000 investors who would receive a letter from me explaining an upcoming change to their investment. If I wrote it with the “WHY” last, like this: 



> “We are stopping your monthly interest payments so the money can be used to improve the properties and thereby increase the value of your investment” I’d get frantic emails and phone calls from about 15% of the audience who freaked out and stopped reading after “We are stopping your monthly interest payments”. Instead, I give them the context-creating WHY at the front of the sentence: “In order to increase the value of the properties, and thereby your investment, we are temporarily pausing your monthly payments.” I hope this example provides the “WHY” for you to remember this tip in all your writing. :)

In Alex's website there are more than 300 pathway images some belong to terpene and others not. Therefore it suited our purpose the best, as we wanted to understnd the statistics behind terpene classification task and learn how difficult is gets for human to classifiy images to narrower domains like [pre- terpene, terpene] from broader domain like [Pathway, Metabolic].



> “as we wanted to understnd the statistics behind terpene classification task and learn how difficult is gets for human to classifiy images” is the WHY in this sentence. I recommend starting with that.

 The table below will give you a glimse of our data looks like for the experimentation.



To find out just how difficult it is for humans to annotate or differentiate whether particular image is of a terpene is a pathway or not, we had circulated 30 images and got 8 annotations collected via GitHub. (Everything was done on open source basis.) 

A simple annotator file looked like this:



The annotated data from the 8 snnotators is then used for the statistical annalysis. You can find discription about the code in here.

Observations:

The annotations analysis gave us some exciting results, from which it was clear that people could differentiate between the basic terminologies. The only difficulty they faced was when the terminologies got specific, like maximum disagreement was in case of terpene



> I don’t know what “maximum disagreement” is a common statistical term that your audience would know, but if so that’s fine.

The result also suggested the modifications for the rules i.e. terpene rules needed framing and sorting so that one could better analyze the image.



> I don’t know what framing and sorting means. To me, it needs explanation or hyperlinks to definitions.

The results also showed that images had maximum non-terpene and pre-terpene pathways.



> Maximum in what way? Why is that significant?



Future Aspects:



> Rather than “Future Aspects”, perhaps “Next Steps”? What title would best describe all the bullet-points you end up including here?



Rules need modification.

Another round of annotations with specific tabel should be circulated.

**Comment by ananyas168** [comment]
## PROJECT 2: Writing a script for capturing Images and Caption from articles over the internet and storing them in a Cproject: 
## Q1. What is the goal?
The goal behind doing this project is to develop a script that automatically extracts and stores images and captions from articles and pushes them into a c-directory.

## Q2 Why are we storing Images and Captions?
 
We are storing images as they are going to be the major interest for the task carried out by  @Roopa and @Anuv. They will be feeding these raw images to OCRs to extract words/phrases.

Whereas the caption text will be used by us to extract words related to terpene. 

The words extracted both the above ways will then be used to build up new dictionaries. Like terpene dictionary etc.

## Steps:
1. To achieve this task we collect articles over the internet using pygetpapers.
2. Parse the XML files provided by pygetpapers
3. Iterate over the XML to capture fig Id:
  3.1 Store the graphic Id 
  3.2 Store the caption text 
4. Dump the caption text in txt format and raw images extracted with the help of graphic id from bin to their respective folders. 


The final C project looks somewhat like this:

![image](https://user-images.githubusercontent.com/66965350/161378530-3a518c9a-6a15-446d-ae79-f6caf8cc73c9.png)

**Reply by EmanuelFaria** [reply]
PROJECT 2: Writing a script for capturing Images and Caption from articles over the internet and storing them in a Cproject:
Q1. What is the goal?
> Replace with “Objective”?
